angular-ngc/defs.bzl

```diff
@@ -0,0 +1,346 @@
+                   "--roots",
+                   native.package_name(),
+                   "%s/%s" % (native.package_name(), html_out),
+               ] +
+               # Generic Assets
+               ["--assets"] + ["$(execpath %s)" % s for s in html_assets] +
+               ["--scripts", "--module", "polyfills-bundle.js"] +
+               # Main bundle to bootstrap the app last
+               ["--scripts", "--module", "%s/main.js" % bundle],
+        # The input HTML template, all assets for potential access for stamping
+        srcs = [":index.html", ":%s" % bundle, ":polyfills-bundle"] + html_assets,
+        visibility = ["//visibility:private"],
+    )
+
+    copy_to_directory(
+        name = name,
+        srcs = [":%s" % bundle, ":polyfills-bundle", ":%s" % html_out] + html_assets + assets,
+        root_paths = [".", "%s/%s" % (native.package_name(), html_out)],
+        visibility = visibility,
+    )
+
+    # http server serving the bundle
+    history_server_bin.history_server_binary(
+        name = "serve" + ("-prod" if production else ""),
+        args = ["$(location :%s)" % name],
+        data = [":%s" % name],
+        visibility = visibility,
+    )
+
+def ng_pkg(name, package_name = None, deps = [], data = [], test_deps = [], visibility = ["//visibility:public"]):
+    """
+    Bazel macro for compiling an npm-like Angular package project. Creates '{name}' and 'test' targets.
+
+    Projects structure:
+      src/
+        public-api.ts
+        **/*.{ts,css,html}
+
+    Tests:
+      src/
+        **/*.spec.ts
+
+    Args:
+      name: the rule name
+      package_name: the package name
+      deps: compilation dependencies
+      data: runtime dependencies
+      test_deps: additional dependencies for tests
+      visibility: visibility of the primary targets ('{name}', 'test')
+    """
+
+    test_spec_srcs = native.glob(["src/**/*.spec.ts"])
+
+    srcs = native.glob(
+        ["src/**/*.ts", "src/**/*.css", "src/**/*.html"],
+        exclude = test_spec_srcs,
+    )
+
+    # An index file to allow direct imports of the directory similar to a package.json "main"
+    write_file(
+        name = "_index",
+        out = "index.ts",
+        content = ["export * from \"./src/public-api\";"],
+        visibility = ["//visibility:private"],
+    )
+
+    ng_project(
+        name = "_lib",
+        srcs = srcs + [":_index"],
+        deps = deps + PACKAGE_DEPS,
+        visibility = ["//visibility:private"],
+    )
+
+    js_library(
+        name = name,
+        srcs = [":_lib"],
+        data = data,
+        visibility = ["//visibility:public"],
+    )
+
+    if package_name != None:
+        npm_package(
+            name = "npm-pkg",
+            package = package_name,
+            srcs = [
+                ":%s" % name,
+            ],
+        )
+
+    if len(test_spec_srcs) > 0:
+        _unit_tests(
+            name = "test",
+            tests = test_spec_srcs,
+            static_files = [],
```

Comment:
Should we add a new parameter `static_files` to `ng_pkg`, `ng_application`?

Reply:
I'm going to just drop all "static file" logic for now since it is unused. No point having something unused in an example...